sigstore-java/src/main/java/dev/sigstore/timestamp/client/TimestampVerifier.java

```diff
@@ -0,0 +1,273 @@
+      throw new TimestampVerificationException("Failed to validate TSA chain", e);
+    }
+
+    // Check if the generation time of the timestamp falls within the validity period of the TSA
+    if (!tsa.getValidFor().contains(tsToken.getTimeStampInfo().getGenTime().toInstant())) {
+      tsaVerificationFailure.put(
+          tsa.getUri().toString(),
+          "Timestamp generation time is not within TSA's validity period.");
+    } else {
+      return;
+    }
+
+    String errors =
+        tsaVerificationFailure.entrySet().stream()
+            .map(entry -> entry.getKey() + " (" + entry.getValue() + ")")
+            .collect(Collectors.joining("\n"));
+    throw new TimestampVerificationException(
+        "Certificate was not verifiable against TSAs\n" + errors);
+  }
+
+  /** Validates the signature of the TimeStampToken using the provided signing certificate. */
+  private void validateTokenSignature(TimeStampToken token, X509Certificate signingCert)
+      throws TimestampVerificationException {
+    try {
+      var verifierBuilder = new JcaSimpleSignerInfoVerifierBuilder();
+      var verifier = verifierBuilder.setProvider("BC").build(signingCert);
+      token.validate(verifier);
+    } catch (OperatorCreationException oce) {
+      throw new TimestampVerificationException("Failed to build SignerInformationVerifier", oce);
+    } catch (TSPException tspe) {
+      throw new TimestampVerificationException("Failed to validate TimeStampToken", tspe);
+    }
+  }
+
+  /** Validates that the provided TSA's certificate chain is self-consistent. */
+  void validateTsaChain(CertificateAuthority tsa, Date tsDate)
+      throws TimestampException,
+          NoSuchProviderException,
+          CertPathValidatorException,
+          InvalidAlgorithmParameterException { // Accept validation date
+    CertPathValidator cpv;
+    try {
+      cpv = CertPathValidator.getInstance("PKIX");
+    } catch (NoSuchAlgorithmException e) {
+      throw new RuntimeException(
+          "No PKIX CertPathValidator, we probably shouldn't be here, but this seems to be a system library error not a program control flow issue",
+          e);
+    }
+
+    PKIXParameters pkixParams;
+    try {
+      pkixParams = new PKIXParameters(Collections.singleton(tsa.asTrustAnchor()));
+    } catch (InvalidAlgorithmParameterException | CertificateException e) {
+      throw new RuntimeException(
+          "Can't create PKIX parameters for the TSA. This should have been checked when generating a verifier instance",
+          e);
+    }
+    pkixParams.setRevocationEnabled(false);
+    pkixParams.setDate(tsDate);
+
+    cpv.validate(tsa.getCertPath(), pkixParams);
+  }
+
+  /**
+   * Finds the TSA that verifies the provided timestamp token by validating its signature using the
+   * certificate chain embedded within the token and matching the leaf to a known TSA.
+   */
+  CertificateAuthority findVerifyingTsaFromEmbeddedCerts(TimeStampToken tsToken)
```

Comment:
Flagging this for more discussion: I wonder if we should drop support for verifying timestamps using embedded certificates, and similarly on the write path, drop support for setting the `certReq` bit.

Embedded certificates are an alternative to distributing a full PKI chain, but you do still need a root of trust. Currently, the TUF metadata assumes you've specified the full chain - you could technically specify just a root, but in practice, a deployment operator should just specify the full chain if they've already specified a root.

cc @loosebazooka

Reply:
Yeah I'm not sure, we kinda landed on this so we could handle all use-cases. But if clients all agree to ignore, I'm fine with that.

Reply:
I've reminded myself what `certReq` actually does, it's for embedding just the TSA leaf certificate in the TimestampResponse. I don't think it's a bad thing to keep this logic in to cover more cases, but if we do, we'll need to change how we find the correct TSA chain in the trust root. Currently, it looks like it's looking for a matching leaf certificate in the TSA chain in the trust root, but I wouldn't expect a chain to contain the leaf cert if it's included in the TSR already. Instead, we'd have to try to verify the full chain, using the CA chain from the trust root and the leaf cert from the TSR.

Reply:
I think this matching is okay?
1. Extract the leaf
2. Match it again a chain in the trusted_root
3. Check the signature
4. Ensure the found TSA has a consistent chain (in validateTsaChain)

Reply:
Rephrasing slightly:

1. Either a) extract the leaf from the Timestamp Response, or b) extract the leaf from the chain in the trusted root
2. Use only the CA certificates in the trusted root to build a pool of intermediate and root certs (sorry, not sure how it's done in java, this is based on Go)
3. Verify the chain from leaf to root (validateTSAChain I think?)
4. Using the public key from the leaf cert, check the signature

Reply:
Sorry yeah, this *is* what we're doing here. I gotta double check the chain verification.
There isn't really a unique in tsa land though. TSA is just providing its certificate, the same leaf will be embedded in all timestamps if requested. This leaf is also provided as part of the chain in our trusted_root.

Reply:
> This leaf is also provided as part of the chain in our trusted_root

This is the part that I was wondering if we should assume. Would we be OK with a leaf certificate not in the trusted root chain?

Reply:
I didn't think so. The leaf is what identifies the TSA, isn't it?

Reply:
Ah I see, you're viewing the embedded leaf as a link to find the correct chain in the trusted root, yea?

I think this is just a different viewpoint, whether you try to verify against every chain in the trust root ([what sigstore-go is doing](https://github.com/sigstore/sigstore-go/blob/cedac1b7e7d4ff3170c12f24f6b53d49a6396c51/pkg/verify/tsa.go#L98) or whether you can find the specific chain with the leaf.

OK, I think both are reasonable approaches.

Reply:
I mean I guess we didn't have to verify the chain of it comes from the trust root.

Reply:
Aaron and I chatted about this - the primary difference between what I was asking about and what's implemented here is just ordering, and I've realized that there is not a duplicated chain verification anywhere. There's no actual risk with doing this in a different order than what I mentioned as well, so this LGTM!